### ISSUE TITLE
sticky needs dependencies that debian doesn't provide (which means debian isn't supported)

### ISSUE THREAD

**Issue by GithubUser699** [linuxmint/sticky#101]
### The Problem
The app installs but throws errors if I want to create a note. Also dpkg states the following messages in the terminal:
```
sticky depends on gir1.2-xapp-1.0 (>= 2.2.0) but only 2.0.7-1 is available.
sticky depends on python3-xapp (>= 2.2.0) but only 2.0.2-2 is available.
```
(errors will be pasted below into "Errors")

### Additional Context
I am using GNOME but tried Cinnamon today. It's different but I like it. Something I especially liked is the sticky app. Months ago I tried to find something similar but had no idea how to search for that... Which is why it made me happy to find sticky! However, I do not intend to switch DEs which is why I wanted to install sticky in GNOME.

### Desired Behavior
It would be great if it would work exactly as it does in linux mint using cinnamon but on my debian install using GNOME instead.

### Things I tried
1. I did everything described [in the debian section of your readme](https://github.com/linuxmint/sticky#mint-debian-ubuntu-and-derivatives) and
2. also [copied the files manually](https://github.com/linuxmint/sticky#otherwise-and-this-is-valid-anywhere-if-you-want-to-avoid-packaging) after the above didn't work.
3. I checked [release 1.0](https://github.com/linuxmint/sticky/releases/tag/1.0) that has the same dependency problems.
4. After 1, 2 and 3 failed I edited the control files (e.g. debian/sticky/DEBIAN/control) and manually inserted the versions that are present in debian 10. It now installed without error or warning, but the same errors I described in "The Problem" show up and the app doesn't work.
5. I tried to convert the app into a flatpak app but since I never did that I wasn't really successful. I don't know how to get all dependencies into the app.

### Errors
Directly after starting the app:
`(sticky.py:1043945): Gtk-CRITICAL **: 15:09:14.391: gtk_widget_get_scale_factor: assertion 'GTK_IS_WIDGET (widget)' failed`

When clicking on "new note":
```
Traceback (most recent call last):
  File "/usr/lib/sticky/sticky.py", line 896, in new_note
    self.add_note(info)
  File "/usr/lib/sticky/sticky.py", line 899, in add_note
    note = self.generate_note(info)
  File "/usr/lib/sticky/sticky.py", line 911, in generate_note
    note = Note(self, self.dummy_window, info)
  File "/usr/lib/sticky/sticky.py", line 155, in __init__
    self.title_style_manager = XApp.StyleManager(widget=self.title_box)
  File "/usr/lib/python3/dist-packages/gi/overrides/__init__.py", line 32, in __getattr__
    return getattr(self._introspection_module, name)
  File "/usr/lib/python3/dist-packages/gi/module.py", line 123, in __getattr__
    raise AttributeError("%r object has no attribute %r" % (
AttributeError: 'gi.repository.XApp' object has no attribute 'StyleManager'
```

When trying to open the settings:
```
Traceback (most recent call last):
  File "/usr/lib/sticky/sticky.py", line 1003, in open_settings_window
    self.settings_window = SettingsWindow(self)
  File "/usr/lib/sticky/sticky.py", line 601, in __init__
    page.pack_start(GSettingsFontButton(_("Font"), SCHEMA, 'font', level=Gtk.FontChooserLevel.SIZE), False, False, 0)
  File "/usr/lib/python3/dist-packages/xapp/GSettingsWidgets.py", line 194, in __init__
    super(NewClass, self).__init__(label, *args, **kwargs)
TypeError: __init__() got an unexpected keyword argument 'level'
```

### System
OS: Debian GNU/Linux 11 (bullseye) 64-bit
GNOME version: 3.38.5
WM: Wayland

**EDIT:**
Also the [runtime dependency versions are wrongly documented in the readme](https://github.com/linuxmint/sticky#runtime-deps) (they are too low compared to the control files)

**Comment by mtwebster** [comment]
Looks like the README is a little out of date. Unfortunately, debian stable doesn't have recent enough package versions to satisfy the dependencies.

You can try the packages here - they're built on bullseye/lmde5- https://github.com/linuxmint/xapp/releases/tag/master.lmde5

Download the packages.tar.gz,  extract only the gir1.2, libxapp1, and xapps-common packages to a folder, then open a terminal there:
```
sudo dpkg -i *.deb 
```

Do the same thing for this package: https://github.com/linuxmint/python3-xapp/releases/tag/master.lmde5.  In this case there's only one package to install.

**Comment by GithubUser699** [comment]
> Looks like the README is a little out of date. Unfortunately, debian stable doesn't have recent enough package versions to satisfy the dependencies.
> 
> You can try the packages here - they're built on bullseye/lmde5- https://github.com/linuxmint/xapp/releases/tag/master.lmde5
> 
> Download the packages.tar.gz, extract only the gir1.2, libxapp1, and xapps-common packages to a folder, then open a terminal there:
> 
> ```
> sudo dpkg -i *.deb 
> ```
> 
> Do the same thing for this package: https://github.com/linuxmint/python3-xapp/releases/tag/master.lmde5. In this case there's only one package to install.

This does kinda work. However,
1.  there is no panel icon which I guess isn't supported (yet?) by sticky in GNOME.
2. The app itself needs to be opened and be present in some workspace, otherwise the notes disappear.

While it's not ideal, it's certainly better than nothing. Thank you for your help!

Do you know btw if installing the files will cause [breakage / FrankenDebian](https://wiki.debian.org/DontBreakDebian#Don.27t_make_a_FrankenDebian)? If it does it would be a huge problem for me.
I might still try to convert it to a flatpak app which shouldn't cause any breakage.

**Comment by collinss** [comment]
>> The app itself needs to be opened and be present in some workspace, otherwise the notes disappear.

> Not sure why this would be - it may be some behavior specific to Gnome-Shell - @collinss ?

I'm not sure I fully understand what you mean, but sticky is designed so that if you don't have a status icon, closing the manager (main window) closes the entire app (along with the notes). This is because the notes need either the status icon or the manager window to control the notes, or quit the app. If this is what you're experiencing, that the behavior is intended. If not, please give me more description of exactly what you're experiencing and what you do to make it happen so that I can look into it more.

**Comment by GithubUser699** [comment]
> There shouldn't be any risk associated with having newer versions of these packages - they were built in a debian bullseye environment, and have no dependencies that aren't part of core debian.

Okay. So far my system seems to work fine too.
> The lack of status icon is due to the fact that gnome-shell doesn't support the types of status icon sticky can currently use. There should be an extension or two that can add that support back, or at least there used to be - Tray-Icons-Reloaded was one.

I tried [Tray Icons: Reloaded](https://extensions.gnome.org/extension/2890/tray-icons-reloaded/) and [TopIcons Plus](https://extensions.gnome.org/extension/1031/topicons/). Both work to add a tray icon for e.g. OBS studio but they won't add something for sticky (OBS needs to be open though. If I close it the tray icon disappears too.)



> I'm not sure I fully understand what you mean, but sticky is designed so that if you don't have a status icon, closing the manager (main window) closes the entire app (along with the notes).

That's exactly what's happening.



Well, it seems like sticky isn't fully supported on debian (probably especially with gnome) which is okay and probably too much to ask of you.

Regarding this ticket: Maybe we can add in the README that we need to do what @mtwebster [mentioned above](https://github.com/linuxmint/sticky/issues/101#issuecomment-1483901693) on debian systems. Also we could update the README with the following version details (as in sticky/debian/control):

> gir1.2-glib-2.0,
> gir1.2-gtk-3.0 (>= 3.22.0),
> gir1.2-xapp-1.0 (>= 2.2.0),
> gir1.2-gspell-1,
> python3,
> python3-gi,
> python3-xapp (>= 2.2.0)
> 

This would resolve this ticket. I could create a PR if you want me to.

**Comment by mtwebster** [comment]
Ok I did some testing on gnome-shell in Ubuntu, and from that initial error message:
```
(sticky.py:1043945): Gtk-CRITICAL **: 15:09:14.391: gtk_widget_get_scale_factor: assertion 'GTK_IS_WIDGET (widget)' failed
```
This tells me you're using wayland. You definitely won't get a tray icon in that case.  If you run in xorg mode, Tray-Icons-Reloaded works fine.

In wayland mode, if I enable the tray icon from sticky's preferences, notes stay open when you close the manager window, but of course there's no way to get the manager window back.  Maybe for now we can detect wayland and add an 'open manager' option to a note's right-click menu if so?

**Comment by GithubUser699** [comment]
> ```
> (sticky.py:1043945): Gtk-CRITICAL **: 15:09:14.391: gtk_widget_get_scale_factor: assertion 'GTK_IS_WIDGET (widget)' failed
> ```
> 
> This tells me you're using wayland. You definitely won't get a tray icon in that case. If you run in xorg mode, Tray-Icons-Reloaded works fine.

I was using wayland, true. How do you now that just from seeing this error message though?
Now that you mentioned it I tested it in xorg too. In that case both tray icon extensions I mentioned earlier show sticky's tray icon.
Also I apparently turned off the tray settings in sticky while testing. So the following I wrote earlier is actually invalid and was only caused by those settings:
>     2. The app itself needs to be opened and be present in some workspace, otherwise the notes disappear.

I should have double checked that, I apologize!

> In wayland mode, if I enable the tray icon from sticky's preferences, notes stay open when you close the manager window, but of course there's no way to get the manager window back.

I can actually just re-launch the sticky app and it will be back. I thought [Launch new instance](https://extensions.gnome.org/extension/600/launch-new-instance/) somehow does that but if I turn off this extension (even if I turn off all extensions) I can still simply re-launch the app. However, if I press alt+TAB I can see the notes and a sticky instance running, even if I close the manager window. Nothing appears if I choose this window though. It's the black window that you can see below:
![sticky_blackWindowAfterClosing](https://user-images.githubusercontent.com/74927253/228321857-e7932d60-bf47-4a21-913d-c05beb5f24d9.png)

**Comment by collinss** [comment]
That black window is a blank off-screen window that I use to tie all the notes together, and show only a single entry in window lists (such as the taskbar and alt-tab) if the appropriate setting is enabled. Check if you have the "Show in taskbar" setting enabled in preferences or not. It's still not working incorrectly either way, because it shouldn't be showing the individual notes in alt-tab ever, but at least that will tell me _how_ it's broken.

I suspect the reason is simply that gnome shell is ignoring the hint that tells it to hide those windows from all window lists, but I can be sure without diving deeper (which unfortunately I don't have time for right now). If that's the case, I may not be able to fix this issue, as there's no way to force it to ignore those windows (at least that I know of).

**Comment by GithubUser699** [comment]
> Check if you have the "Show in taskbar" setting enabled in preferences or not.

It was turned off, but nothing changes if I turn it on / off.  Also, if I run sticky in the terminal and change this setting, nothing (no error, warning...) appears.

> I suspect the reason is simply that gnome shell is ignoring the hint that tells it to hide those windows from all window lists, but I can be sure without diving deeper (which unfortunately I don't have time for right now). If that's the case, I may not be able to fix this issue, as there's no way to force it to ignore those windows (at least that I know of).

That's alright, no worries! You guys helped me to run sticky and I'm grateful for that.

**Comment by Miccio-AN** [comment]
Just tried to compile and install on Debian 12 Bookworm:

1) on X11 everything works perfectly
2) on Wayland the icon does not appear in the tray but the software seems to work well

Maybe it can help someone know that all dependencies are already present in Debian 12

**Comment by RENANZG** [comment]
Noob here. Is it safe to install the .deb file by the command "apt install -f " on Debian 12 ? Won't I have problems with broken dependencies?

**Comment by collinss** [comment]
@RENANZG Assuming we have the dependencies set up correctly, the .deb wont install normailly if you don't have all the dependencies, so it should be safe to install it if you use a tool like dpkg or gdebi to install it.

**Comment by Miccio-AN** [comment]
Always talking about Debian 12 and Wayland. 
Given, as I wrote above, that with X11 (and Debian 12) everything works perfectly.

I noticed this: by starting sticky from the terminal with root permissions (I know it shouldn't be done) everything works as it should (the icon appears in the tray ... etc ...). Except, of course, on the next reboot.

I write this because maybe it can be useful to the developer.
Thanks!
